File: 57.插入区间.go
/*
 * @lc app=leetcode.cn id=57 lang=golang
 *
 * [57] 插入区间
 */

// @lc code=start
package main
func insert(intervals [][]int, newInterval []int) [][]int {
	left,right := newInterval[0],newInterval[1]
	merged := false
	ans := [][]int{}
	min := func(x,y int) int {
		if x < y {
			return x
		}
		return y
	}
	max := func(x,y int) int {
		if x > y {
			return x
		}
		return y
	}
	for _,interval := range intervals {
		if interval[0] > right {
			if !merged {
				// ans = append(ans,newInterval)
				// 不写上面这句是因为可能已经经过合并了，left和right更新
				ans = append(ans,[]int{left,right})
				merged = true
			}
			ans = append(ans,interval)
		}else if interval[1] < left {
			ans = append(ans,interval)
		}else{
			// 有交集就计算并集
			left = min(left,interval[0])
			right = max(right,interval[1])
		}
	}
	if !merged {
		ans = append(ans,[]int{left,right})
	}
	return ans
}

// @lc code=end

File: 57.插入区间.py
#
# @lc app=leetcode.cn id=57 lang=python3
#
# [57] 插入区间
#

# @lc code=start
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        result: list[list[int]] = []
        l, r = newInterval

        for i, interval in enumerate(intervals):
            if interval[1] < l:
                result.append(interval)
            elif interval[0] > r:
                result.append([l,r])
                result.extend(intervals[i:])
                return result
            else:
                l, r = min(l, interval[0]), max(r, interval[1])
        
        result.append([l, r])
        return result
# @lc code=end

